src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,551 @@
+		if (px || py || pz)
+			return false;
+	}
+
+	return true;
+}
+
+int tdfb_direction_init(struct tdfb_comp_data *cd, int32_t fs, int ch_count)
+{
+	struct sof_eq_iir_header_df2t *filt;
+	int64_t *delay;
+	int32_t d_max;
+	int32_t t_max;
+	size_t size;
+	int n;
+	int i;
+
+	/* Select emphasis response per sample rate */
+	switch (fs) {
+	case 16000:
+		filt = (struct sof_eq_iir_header_df2t *)iir_emphasis_16k;
+		break;
+	case 48000:
+		filt = (struct sof_eq_iir_header_df2t *)iir_emphasis_48k;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* Allocate delay lines for IIR filters and initialize them */
+	size = ch_count * iir_delay_size_df2t(filt);
+	delay = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, size);
+	if (!delay)
+		return -EINVAL;
+
+	cd->direction.df2t_delay = delay;
+	for (i = 0; i < ch_count; i++) {
+		iir_init_coef_df2t(&cd->direction.emphasis[i], filt);
+		iir_init_delay_df2t(&cd->direction.emphasis[i], &delay);
+	}
+
+	/* Unit delay length in Q1.31 seconds */
+	cd->direction.unit_delay = ((int64_t)1 << 31) / fs;
+
+	/* Get max possible mic-mic distance, then from max distance to max time t = d_max / v,
+	 * t is Q1.15, d is Q4.12, v is Q9.0
+	 */
+	d_max = max_mic_distance(cd);
+	t_max = (d_max << 3) / SPEED_OF_SOUND;
+
+	/* Calculate max lag to search and allocate delay line for cross correlation
+	 * compute. Add one to make sure max possible lag is in search window.
+	 */
+	cd->direction.max_lag = Q_MULTSR_32X32(fs, t_max, 0, 15, 0) + 1;
+	n = (cd->max_frames + (2 * cd->direction.max_lag + 1)) * ch_count;
+	cd->direction.d_size =  n * sizeof(int16_t);
+	cd->direction.d = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, cd->direction.d_size);
+	if (!cd->direction.d) {
+		rfree(cd->direction.df2t_delay);
+		return -EINVAL;
+	}
+
+	/* Set needed pointers to xcorr delay line, advance write pointer by max_lag to keep read
+	 * always behind write
+	 */
+	cd->direction.d_end = cd->direction.d + n;
+	cd->direction.wp = cd->direction.d;
+	cd->direction.rp = cd->direction.wp;
+	cd->direction.wp += ch_count * (cd->direction.max_lag + 1);
+
+	/* xcorr result is temporary but too large for stack so it is allocated here */
+	cd->direction.r_size = (2 * cd->direction.max_lag + 1) * sizeof(int32_t);
+	cd->direction.r = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, cd->direction.r_size);
+	if (!cd->direction.r) {
+		rfree(cd->direction.df2t_delay);
+		rfree(cd->direction.d);
+		return -EINVAL;
+	}
+
+	/* Check for line array mode */
+	cd->direction.line_array = line_array_mode_check(cd);
+
+	/* Initialize direction to zero radians, set initial step sign to +1 */
+	cd->direction.az = 0;
+	cd->direction.step_sign = 1;
+	return 0;
+}
+
+void tdfb_direction_free(struct tdfb_comp_data *cd)
+{
+	rfree(cd->direction.df2t_delay);
+	rfree(cd->direction.d);
+	rfree(cd->direction.r);
+}
+
+/* Measure level of one channel */
+static inline void level_update(struct tdfb_comp_data *cd, int frames, int ch_count, int channel)
```

Comment:
these are some big functions to inline, what is the reasoning?

Reply:
I was thinking speed, but I don't think there's a difference. I'll change to no inline.